src/coreclr/jit/lower.cpp

```diff
@@ -8030,27 +8030,302 @@ void Lowering::MapParameterRegisterLocals()
                 comp->m_paramRegLocalMappings->Emplace(&segment, fieldLclNum, segment.Offset - fieldDsc->lvFldOffset);
             }
 
-            LclVarDsc* fieldLclDsc = comp->lvaGetDesc(fieldLclNum);
-            assert(!fieldLclDsc->lvIsParamRegTarget);
-            fieldLclDsc->lvIsParamRegTarget = true;
+            assert(!fieldDsc->lvIsParamRegTarget);
+            fieldDsc->lvIsParamRegTarget = true;
         }
     }
 
+    if (!comp->opts.IsOSR())
+    {
+        FindInducedParameterRegisterLocals();
+    }
+
 #ifdef DEBUG
     if (comp->verbose)
     {
         printf("%d parameter register to local mappings\n", comp->m_paramRegLocalMappings->Height());
         for (int i = 0; i < comp->m_paramRegLocalMappings->Height(); i++)
         {
             const ParameterRegisterLocalMapping& mapping = comp->m_paramRegLocalMappings->BottomRef(i);
-            printf("  ");
-            mapping.RegisterSegment->Dump();
-            printf(" -> V%02u\n", mapping.LclNum);
+            printf("  %s -> V%02u+%u\n", getRegName(mapping.RegisterSegment->GetRegister()), mapping.LclNum,
+                   mapping.Offset);
         }
     }
 #endif
 }
 
+//------------------------------------------------------------------------
+// Lowering::FindInducedParameterRegisterLocals:
+//   Find locals that would be profitable to map from parameter registers,
+//   based on IR in the initialization block.
+//
+void Lowering::FindInducedParameterRegisterLocals()
+{
+#ifdef TARGET_ARM
+    // On arm32 the profiler hook does not preserve arg registers, so
+    // parameters are prespilled and cannot stay enregistered.
+    if (comp->compIsProfilerHookNeeded())
+    {
+        JITDUMP("Skipping FindInducedParameterRegisterLocals on arm32 with profiler hook\n");
+        return;
+    }
+#endif
+
+    // Check if we possibly have any parameters we can induce new register
+    // locals from.
+    bool anyCandidates = false;
+    for (unsigned lclNum = 0; lclNum < comp->info.compArgsCount; lclNum++)
+    {
+        LclVarDsc* lcl = comp->lvaGetDesc(lclNum);
+        if (lcl->lvPromoted || !lcl->lvDoNotEnregister)
+        {
+            continue;
+        }
+
+        const ABIPassingInformation& abiInfo = comp->lvaGetParameterABIInfo(lclNum);
+        if (!abiInfo.HasAnyRegisterSegment())
+        {
+            continue;
+        }
+
+        anyCandidates = true;
+        break;
+    }
+
+    if (!anyCandidates)
+    {
+        return;
+    }
+
+    bool     hasRegisterKill = false;
+    LocalSet storedToLocals(comp->getAllocator(CMK_ABI));
+    // Now look for optimization opportunities in the first block: places where
+    // we read fields out of struct parameters that can be mapped cleanly. This
+    // is frequently created by physical promotion.
+    for (GenTree* node : LIR::AsRange(comp->fgFirstBB))
+    {
+        hasRegisterKill |= node->IsCall();
```

Comment:
>hasRegisterKill [](http://example.com/codeflow?start=8&length=15)

are there stores that are outside of struct parameters copying that we should not be visiting in first block? Or are the stores always guaranteed to be coming from struct params?

Reply:
There can be any other stores, but the checks below should filter the ones that aren't relevant out